Web Inspector: Adopt esprima-next fork of esprima to support ES2022 syntax

patrickangle opened the pull request:
#### 2c7d4e7fbcc5794d397898df5de15f8d9faebd29
<pre>
Web Inspector: Adopt esprima-next fork of esprima to support ES2022 syntax
<a href="https://bugs.webkit.org/show_bug.cgi?id=253856">https://bugs.webkit.org/show_bug.cgi?id=253856</a>
rdar://89097522

Reviewed by Devin Rousso.

Adopt a fork of esprima, esprima-next, which has been updated with ES2022 features, and also includes our previously
up-streamed changes, as well as equivalent fixes for patches that had not been up-streamed. This commit includes commit
a002247e9cb92768ea91ef838b223012fe4fee9f of <a href="https://github.com/node-projects/esprima-next/.">https://github.com/node-projects/esprima-next/.</a>

Some adoption on JSFormatter/ESTreeWalker is necessary to handle a renamed piece of syntax, as well as support both
public and private class-level variable declarations.

* LayoutTests/inspector/formatting/formatting-javascript-expected.txt:
* LayoutTests/inspector/formatting/formatting-javascript.html:
- Fix simple class test case to extend some superclass to allow super to legally be used.
- Add cases to verify ImportExpression, ChainExpression, PrivateIdentifier, ImportAttribute, and StaticBlock types.

* LayoutTests/inspector/formatting/formatting-javascript-expected.txt:
* LayoutTests/inspector/formatting/formatting-javascript.html:

* LayoutTests/inspector/formatting/resources/javascript-tests/classes-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/classes.js:
- Add new test case for static blocks.

* LayoutTests/inspector/formatting/resources/javascript-tests/modules-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/modules.js:
- Add new test cases for conditional assignments.

* LayoutTests/inspector/formatting/resources/javascript-tests/optional-chaining-expected.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/optional-chaining.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/private-classes-expected.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/private-classes.js: Added.
- Add new test cases for optional chaining and private class members.

* LayoutTests/inspector/formatting/resources/javascript-tests/unary-binary-expressions-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/unary-binary-expressions.js:
- Add new test cases import assertions.

* Source/WebInspectorUI/Tools/JSFormatter/index.html:

* Source/WebInspectorUI/UserInterface/External/Esprima/esprima.js:
- Updated to <a href="https://github.com/node-projects/esprima-next/commit/a002247e9cb92768ea91ef838b223012fe4fee9f">https://github.com/node-projects/esprima-next/commit/a002247e9cb92768ea91ef838b223012fe4fee9f</a>

* Source/WebInspectorUI/UserInterface/Models/ScriptSyntaxTree.js:
(WI.ScriptSyntaxTree.prototype._recurse):
- Handle new types from esprima and convert them into our AST tree, based on the estree spec: <a href="https://github.com/estree/estree">https://github.com/estree/estree</a>

* Source/WebInspectorUI/UserInterface/Workers/Formatter/ESTreeWalker.js:
(ESTreeWalker.prototype._walkChildren):
(ESTreeWalker):
- Add pass-through cases for optional chaining and private identifiers.
- Adopt renamed &quot;ImportExpression&quot; (previously &quot;Import&quot;).

* Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
(JSFormatter.prototype._before):
(JSFormatter.prototype._after):
- ChainExpression tokens don&apos;t have a range.

(JSFormatter.prototype._handleTokenAtNode):
- Added handling of &quot;PrivateIdentifier&quot;, matching existing &quot;Identifier&quot; handling.
- Fixed issue where generator functions would incorrectly attempt to remove the previous white space before a &quot;*&quot;, even
if there wasn&apos;t any white space to remove.
- Add support for static properties, private properties, and property assignment inside classes.
- Add support for private methods.
- Add generic handling of &quot;AssignmentPattern&quot; and &quot;ChainExpression&quot;.

Canonical link: <a href="https://commits.webkit.org/261748@main">https://commits.webkit.org/261748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e8dd3d1f20dd7388fbec34b274173f7a390b1a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112630 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21784 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1295 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4406 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121166 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116697 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5557 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118398 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17174 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100400 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105705 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99127 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/946 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46193 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14114 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1011 "6 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/95298 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14795 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10340 "2 flakes 2 failures") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20132 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52980 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8197 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16637 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/ESTreeWalker.js:
```diff
@@ -288,11 +288,13 @@ ESTreeWalker = class ESTreeWalker
         case "DebuggerStatement":
         case "EmptyStatement":
         case "Identifier":
-        case "Import":
```
You'll want to update `ScriptSyntaxTree.js` with the new node types and ensure that the AST is walkable. If we're removing `Import` we probably need to add `ImportExpression`, `ChainExpression`, and `PrivateIdentifier`.

* `WI.ScriptSyntaxTree.NodeType ` enum should have new types <https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Models/ScriptSyntaxTree.js#L1054>
* `WI.ScriptSyntaxTree.prototype._recurse` to handle the new types <https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Models/ScriptSyntaxTree.js#L310>
* `WI.ScriptSyntaxTree._createInternalSyntaxTree` to handle the new types.

patrickangle replied:
Thank you for bringing these up. This is now handled and tested in the latest PR.

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
```diff
@@ -463,6 +467,20 @@ JSFormatter = class JSFormatter
             return;
         }
 
+        if (nodeType === "ImportExpression") {
```
Do you need to update the `import.js` test for these changes?

patrickangle replied:
This actually isn't necessary in my upcoming PR, which actually correctly handles ImportExpression from esprima all the way to our ast tree.

dcrousso commented on LayoutTests/inspector/formatting/resources/javascript-tests/optional-chaining.js:
```diff
@@ -0,0 +1,3 @@
+foo?.val?.prop;
```
can/should we have these be a bit more crazy?  e.g. `foo  ?.  val  ?.  prop  ;` to really make sure formatting is actually happening

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/ScriptSyntaxTree.js:
```diff
@@ -814,6 +823,12 @@ WI.ScriptSyntaxTree = class ScriptSyntaxTree
+        case "PrivateIdentifier":
+            result = {
+                type: WI.ScriptSyntaxTree.NodeType.PrivateIdentifier,
+                name: node.name
```
Style: missing `,`

dcrousso commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
```diff
@@ -161,7 +161,7 @@ JSFormatter = class JSFormatter
     _before(node)
     {
-        if (!node.parent)
+        if (!node.parent || !node.range)
```
NIT: should we move these checks to `ESTreeWalker` instead so we dont have to have them in every `_before`/`_after`?

patrickangle replied:
We could, although that feels like it causes this implementation detail of JSFormatter to leak into ESTreeWalker. This is a case where we are making a decision based on what properties the node has what to do, which includes what to do if a range is not present.

dcrousso commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
```diff
@@ -444,7 +447,8 @@ JSFormatter = class JSFormatter
                 if (tokenValue === "*") {
-                    builder.removeLastWhitespace();
+                    if (builder.lastTokenWasWhitespace)
+                        builder.removeLastWhitespace();
```
NIT: is this only because of the `console.assert` in `FormatterContentBuilder.prototype.removeLastWhitespace`?  IMO we should consider maybe removing it and renaming to `tryRemoveLastWhitespace` or somethin since it seems much more useful as a "do it if you can" rather than a "you must be in this state"

dcrousso commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
```diff
@@ -707,15 +711,19 @@ JSFormatter = class JSFormatter
 
         if (nodeType === "Property") {
-            console.assert(tokenValue === ":" || tokenValue === "get" || tokenValue === "set" || tokenValue === "async" || tokenValue === "*" || tokenValue === "[" || tokenValue === "]" || tokenValue === "(" || tokenValue === ")", token);
+            console.assert(tokenValue === ":" || tokenValue === "static" || tokenValue === "get" || tokenValue === "set" || tokenValue === "async" || tokenValue === "*" || tokenValue === "[" || tokenValue === "]" || tokenValue === "(" || tokenValue === ")" || tokenValue === "#" || tokenValue === "=", token);
```
NIT: maybe change this to a `[":", "static", ...].includes(tokenValue)` to make it less verbose?

dcrousso reviewed (approved):
i trust you've already done your homework, but please double check to make sure that you've covered _all_ of the new tokens/types :)

super excited to be able to prettify inspector^2 again! 🎉

patrickangle committed:
Web Inspector: Adopt esprima-next fork of esprima to support ES2022 syntax
https://bugs.webkit.org/show_bug.cgi?id=253856
rdar://89097522

Reviewed by Devin Rousso.

Adopt a fork of esprima, esprima-next, which has been updated with ES2022 features, and also includes our previously
up-streamed changes, as well as equivalent fixes for patches that had not been up-streamed. This commit includes commit
a002247e9cb92768ea91ef838b223012fe4fee9f of https://github.com/node-projects/esprima-next/.

Some adoption on JSFormatter/ESTreeWalker is necessary to handle a renamed piece of syntax, as well as support both
public and private class-level variable declarations.

* LayoutTests/inspector/formatting/formatting-javascript-expected.txt:
* LayoutTests/inspector/formatting/formatting-javascript.html:
- Fix simple class test case to extend some superclass to allow super to legally be used.
- Add cases to verify ImportExpression, ChainExpression, PrivateIdentifier, ImportAttribute, and StaticBlock types.

* LayoutTests/inspector/formatting/formatting-javascript-expected.txt:
* LayoutTests/inspector/formatting/formatting-javascript.html:

* LayoutTests/inspector/formatting/resources/javascript-tests/classes-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/classes.js:
- Add new test case for static blocks.

* LayoutTests/inspector/formatting/resources/javascript-tests/modules-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/modules.js:
- Add new test cases for conditional assignments.

* LayoutTests/inspector/formatting/resources/javascript-tests/optional-chaining-expected.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/optional-chaining.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/private-classes-expected.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/private-classes.js: Added.
- Add new test cases for optional chaining and private class members.

* LayoutTests/inspector/formatting/resources/javascript-tests/unary-binary-expressions-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/unary-binary-expressions.js:
- Add new test cases import assertions.

* Source/WebInspectorUI/Tools/JSFormatter/index.html:

* Source/WebInspectorUI/UserInterface/External/Esprima/esprima.js:
- Updated to https://github.com/node-projects/esprima-next/commit/a002247e9cb92768ea91ef838b223012fe4fee9f

* Source/WebInspectorUI/UserInterface/Models/ScriptSyntaxTree.js:
(WI.ScriptSyntaxTree.prototype._recurse):
- Handle new types from esprima and convert them into our AST tree, based on the estree spec: https://github.com/estree/estree

* Source/WebInspectorUI/UserInterface/Workers/Formatter/ESTreeWalker.js:
(ESTreeWalker.prototype._walkChildren):
(ESTreeWalker):
- Add pass-through cases for optional chaining and private identifiers.
- Adopt renamed "ImportExpression" (previously "Import").

* Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
(JSFormatter.prototype._before):
(JSFormatter.prototype._after):
- ChainExpression tokens don't have a range.

(JSFormatter.prototype._handleTokenAtNode):
- Added handling of "PrivateIdentifier", matching existing "Identifier" handling.
- Fixed issue where generator functions would incorrectly attempt to remove the previous white space before a "*", even
if there wasn't any white space to remove.
- Add support for static properties, private properties, and property assignment inside classes.
- Add support for private methods.
- Add generic handling of "AssignmentPattern" and "ChainExpression".

Canonical link: https://commits.webkit.org/261748@main
